src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_battery_power_formula.py

```diff
@@ -65,39 +69,106 @@ def generate(
 
         component_graph = connection_manager.get().component_graph
 
-        battery_inverters = frozenset(
-            frozenset(
+        # all_connected_batteries has to be set because
+        # multiple inverters can be connected to the same battery
+        inv_bat_mapping: dict[Component, set[Component]] = {}
+        all_connected_batteries: set[Component] = set()
+        for bat_id in component_ids:
+            inverters = set(
                 filter(
                     component_graph.is_battery_inverter,
                     component_graph.predecessors(bat_id),
                 )
             )
-            for bat_id in component_ids
-        )
-
-        if not all(battery_inverters):
-            raise ComponentNotFound(
-                "All batteries must have at least one inverter as a predecessor."
-            )
+            if len(inverters) == 0:
+                raise ComponentNotFound(
+                    "All batteries must have at least one inverter as a predecessor.",
+                    "Battery ID %s has no inverter as a predecessor.",
+                    bat_id,
+                )
 
-        all_connected_batteries = set()
-        for inverters in battery_inverters:
             for inverter in inverters:
-                all_connected_batteries.update(
-                    component_graph.successors(inverter.component_id)
+                inv_bat_mapping[inverter] = component_graph.successors(
+                    inverter.component_id
                 )
+                all_connected_batteries.update(inv_bat_mapping[inverter])
 
         if len(all_connected_batteries) != len(component_ids):
             raise FormulaGenerationError(
                 "All batteries behind a set of inverters must be requested."
             )
```

Comment:
This looks wrong as well,  just checking the lengths is not enough, we should check that `all_connected_batteries` is a subset of `component_ids`.

But because this is an older issue, we can do it separately if you don't want to touch it now.

Reply:
Fixed